### DISCUSSION TITLE
How to show only selected rows in AG Grid?

### DISCUSSION THREAD

**Discussion by fmcdevelopment** [zauberzeug/nicegui#3371] [Q&A] [answered]
### Question

I am trying to implement a switch that will trigger AG Grid to show only selected rows. inspired by [this](https://github.com/ag-grid/ag-grid/issues/1312#issuecomment-509651169), I wrote something like below:

```py
from nicegui import ui

async def show_selected_rows(grid_target:ui.aggrid, show_selected_rows:bool=False) -> None:
    if show_selected_rows:
        grid_target.options[':isExternalFilterPresent'] = "() => true"
        grid_target.run_grid_method('onFilterChanged')

ui.switch('Show selected rows', on_change=lambda e: show_selected_rows(grid_target=grid, show_selected_rows=e.value))
grid = ui.aggrid({
    'defaultColDef': {'flex': 1},
    'columnDefs': [
        {'headerName': 'Name', 'field': 'name'},
        {'headerName': 'Age', 'field': 'age'},
        {'headerName': 'Parent', 'field': 'parent'},
    ],
    'rowData': [
        {'name': 'Alice', 'age': 18, 'parent': 'David'},
        {'name': 'Bob', 'age': 21, 'parent': 'Eve'},
        {'name': 'Carol', 'age': 42, 'parent': 'Frank'},
    ],
    'rowSelection': 'multiple',
    ':isExternalFilterPresent': '() => false',
    ':doesExternalFilterPass': "(node) => node.isSelected()",
}).classes('max-h-40')

ui.run()
```

In the official AG Grid documentation, it is possible to control an external filter shown [here](https://www.ag-grid.com/javascript-data-grid/filter-external/#example). I have also seen [#2843](https://github.com/zauberzeug/nicegui/discussions/2843) but I want to control the filter to be ON and OFF using the switch. Is this possible? I am new to NiceGUI, so sorry if this does not make sense. 
Thank you.

**Comment by falkoschindler** [comment, marked as answer]
Hi @fmcdevelopment,

The problem seems to be that the `grid` needs to be updated after changing the "isExternalFilterPresent" option. But calling `grid.update()` deletes the current row selection.

A simple workaround might be to leave the "isExternalFilterPresent" option unchanged and to use a global flag `window.showSelectedRows` instead:

```py
def show_selected_rows(show: bool) -> None:
    ui.run_javascript(f'window.showSelectedRows = {"true" if show else "false"};')
    grid.run_grid_method('onFilterChanged')

ui.switch('Show selected rows', on_change=lambda e: show_selected_rows(e.value))

grid = ui.aggrid({
    'columnDefs': [{'headerName': 'Name', 'field': 'name'}],
    'rowData': [{'name': name} for name in ['Alice', 'Bob', 'Carol']],
    'rowSelection': 'multiple',
    ':isExternalFilterPresent': '() => window.showSelectedRows',
    ':doesExternalFilterPass': '(node) => node.isSelected()',
})
```

**Comment by fmcdevelopment** [comment]
Hi @falkoschindler,

Awesome, it works perfectly. Thank you. 
I also looked at your answer [here](https://github.com/zauberzeug/nicegui/discussions/3182#discussioncomment-9839755) where you pointed out the possibility of using `setGridOption` but I ran into the same error as you wrote. I will be waiting for `setGridOption` to be available as it will give more power to the NiceGUI user trying to leverage AG Grid capabilities.
